File: src/forecast.py
"""Recursive multi-step forecasting built on top of the single-step LSTM.

The trained model predicts exactly 1 hour ahead from a 24-hour lookback
window. To forecast an arbitrary horizon ("any hour"), this module feeds
each prediction back in as if it were an observed value, recomputing lag
and rolling features from a hybrid actual+predicted history so later steps
never see true future electricity — only what a real deployment would have
at that point (its own prior forecasts).

Weather and calendar features for the forecast window are taken from the
historical record, standing in for a live weather-forecast feed (this
dataset has no forecast weather). That means valid forecasts are limited to
timestamps inside the historical range — this module doubles as an
interactive backtesting tool as well as a forecaster.
"""
from __future__ import annotations

import json

import joblib
import numpy as np
import pandas as pd

from src import config

_cache: dict = {}


def _load_artifacts():
    if not _cache:
        from tensorflow import keras

        _cache["model"] = keras.models.load_model(config.MODEL_PATH)
        _cache["feature_scaler"] = joblib.load(config.FEATURE_SCALER_PATH)
        _cache["target_scaler"] = joblib.load(config.TARGET_SCALER_PATH)
        with open(config.FEATURE_LIST_PATH) as f:
            _cache["feature_cols"] = json.load(f)
        _cache["cleaned"] = pd.read_csv(config.CLEANED_DATA_PATH, index_col="timestamp", parse_dates=True)
        _cache["featured"] = pd.read_csv(config.FEATURED_DATA_PATH, index_col="timestamp", parse_dates=True)
    return _cache


def _time_features(ts: pd.Timestamp) -> dict:
    hour, dow, month = ts.hour, ts.dayofweek, ts.month
    return {
        "hour_sin": np.sin(2 * np.pi * hour / 24),
        "hour_cos": np.cos(2 * np.pi * hour / 24),
        "dow_sin": np.sin(2 * np.pi * dow / 7),
        "dow_cos": np.cos(2 * np.pi * dow / 7),
        "month_sin": np.sin(2 * np.pi * month / 12),
        "month_cos": np.cos(2 * np.pi * month / 12),
        "is_weekend": 1 if dow >= 5 else 0,
    }


def get_bounds(max_horizon: int = 72):
    """Return (min_start, max_end, max_horizon): the range of timestamps a
    forecast can legally start/end at, given the historical data on hand."""
    art = _load_artifacts()
    min_start = art["featured"].index.min() + pd.Timedelta(hours=config.LOOKBACK)
    max_end = art["cleaned"].index.max()
    return min_start, max_end, max_horizon


def recursive_forecast(start: str | pd.Timestamp, horizon: int) -> list[dict]:
    art = _load_artifacts()
    model = art["model"]
    feature_scaler = art["feature_scaler"]
    target_scaler = art["target_scaler"]
    feature_cols = art["feature_cols"]
    cleaned = art["cleaned"]
    featured = art["featured"]

    start_ts = pd.Timestamp(start)
    min_start = featured.index.min() + pd.Timedelta(hours=config.LOOKBACK)
    max_end = cleaned.index.max()

    if start_ts < min_start or start_ts > max_end:
        raise ValueError(f"start must be between {min_start} and {max_end}")

    end_ts = start_ts + pd.Timedelta(hours=horizon - 1)
    if end_ts > max_end:
        max_horizon_here = int((max_end - start_ts) / pd.Timedelta(hours=1)) + 1
        raise ValueError(
            f"horizon too large for this start time — historical data (used as the weather-forecast "
            f"stand-in) only extends to {max_end}. Max horizon from this start is {max_horizon_here}h."
        )

    elec_actual = cleaned[config.TARGET_COL]
    predicted: dict[pd.Timestamp, float] = {}
    feat_cache: dict[pd.Timestamp, np.ndarray] = {}

    def elec_value(ts: pd.Timestamp) -> float:
        if ts in predicted:
            return predicted[ts]
        return float(elec_actual.loc[ts])

    def get_feature_row(ts: pd.Timestamp) -> np.ndarray:
        if ts in feat_cache:
            return feat_cache[ts]
        if ts < start_ts:
            row = featured.loc[ts, feature_cols].to_numpy(dtype=float)
            feat_cache[ts] = row
            return row

        row_dict = {c: float(cleaned.loc[ts, c]) for c in config.WEATHER_FEATURE_COLS}
        row_dict.update(_time_features(ts))
        for lag in config.LAG_HOURS:
            row_dict[f"{config.TARGET_COL}_lag_{lag}"] = elec_value(ts - pd.Timedelta(hours=lag))
        for w in config.ROLLING_WINDOWS:
            vals = [elec_value(ts - pd.Timedelta(hours=k)) for k in range(1, w + 1)]
            row_dict[f"{config.TARGET_COL}_roll_mean_{w}"] = float(np.mean(vals))
            row_dict[f"{config.TARGET_COL}_roll_std_{w}"] = float(np.std(vals, ddof=1))

        row = np.array([row_dict[c] for c in feature_cols], dtype=float)
        feat_cache[ts] = row
        return row

    results = []
    for h in range(horizon):
        target_ts = start_ts + pd.Timedelta(hours=h)
        window_ts = [target_ts - pd.Timedelta(hours=k) for k in range(config.LOOKBACK, 0, -1)]
        window_feats = pd.DataFrame(np.stack([get_feature_row(t) for t in window_ts]), columns=feature_cols)
        scaled = feature_scaler.transform(window_feats)
        X = scaled[np.newaxis, :, :]
        y_scaled = float(np.asarray(model(X, training=False)).ravel()[0])
        y_pred = float(target_scaler.inverse_transform([[y_scaled]])[0, 0])
        predicted[target_ts] = y_pred

        actual_val = float(elec_actual.loc[target_ts]) if target_ts in elec_actual.index else None
        results.append({"timestamp": target_ts.isoformat(), "predicted": round(y_pred, 4), "actual": actual_val})

    return results


if __name__ == "__main__":
    min_start, max_end, max_h = get_bounds()
    demo_start = max_end - pd.Timedelta(hours=48)
    print(f"valid range: {min_start} -> {max_end}")
    out = recursive_forecast(demo_start, horizon=24)
    for row in out[:5]:
        print(row)
    print("...")
    errs = [abs(r["predicted"] - r["actual"]) for r in out if r["actual"] is not None]
    print(f"MAE over demo window: {np.mean(errs):.4f}")

File: docs/js/forecast.js
// Client-side port of src/forecast.py — recursive multi-hour forecasting.
// Runs entirely in the browser against the static history.json + the
// TensorFlow.js model, so it works with no backend (GitHub Pages).
//
// Same approach as the Python version: to forecast beyond 1 hour, each
// prediction is fed back in as if it were an observed value, so later
// steps' lag/rolling features are built from a hybrid actual+predicted
// history — never from true future electricity. Weather/calendar features
// for the forecast window come from the historical record (no live
// weather feed), so valid forecasts are limited to the historical period.

const ForecastEngine = (function () {
  let state = null; // { history, meta, scalers, model }

  function hourMs(n) { return n * 3600 * 1000; }

  function toHourTs(d) {
    const t = new Date(d);
    t.setMinutes(0, 0, 0);
    return t;
  }

  function isoHour(d) {
    // "YYYY-MM-DDTHH:00:00" in local wall-clock time (matches the dataset,
    // which has no timezone info — it's treated as naive local hours throughout).
    const pad = (n) => String(n).padStart(2, '0');
    return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T${pad(d.getHours())}:00:00`;
  }

  function timeFeatures(d) {
    const hour = d.getHours();
    const dow = (d.getDay() + 6) % 7; // JS: 0=Sun..6=Sat -> convert to Python's 0=Mon..6=Sun
    const month = d.getMonth() + 1;
    return {
      hour_sin: Math.sin((2 * Math.PI * hour) / 24),
      hour_cos: Math.cos((2 * Math.PI * hour) / 24),
      dow_sin: Math.sin((2 * Math.PI * dow) / 7),
      dow_cos: Math.cos((2 * Math.PI * dow) / 7),
      month_sin: Math.sin((2 * Math.PI * month) / 12),
      month_cos: Math.cos((2 * Math.PI * month) / 12),
      is_weekend: dow >= 5 ? 1 : 0,
    };
  }

  async function init() {
    if (state) return state;
    const [history, meta, scalers, model] = await Promise.all([
      fetch(assetUrl('data/history.json')).then((r) => r.json()),
      fetch(assetUrl('data/meta.json')).then((r) => r.json()),
      fetch(assetUrl('data/scalers.json')).then((r) => r.json()),
      loadTfjsModel(assetUrl('model/weights.json')),
    ]);

    const elecByTs = new Map();
    const weatherByTs = new Map();
    for (let i = 0; i < history.timestamps.length; i++) {
      const ts = history.timestamps[i];
      elecByTs.set(ts, history[meta.target_col][i]);
      const w = {};
      for (const col of meta.weather_cols) w[col] = history[col][i];
      weatherByTs.set(ts, w);
    }

    const minTs = new Date(history.timestamps[0]);
    const maxTs = new Date(history.timestamps[history.timestamps.length - 1]);
    const minStart = new Date(minTs.getTime() + hourMs(meta.lookback + Math.max(...meta.lag_hours)));

    state = { meta, scalers, model, elecByTs, weatherByTs, minTs, maxTs, minStart };
    return state;
  }

  function getBounds() {
    if (!state) throw new Error('ForecastEngine not initialized');
    return { minStart: state.minStart, maxEnd: state.maxTs, maxHorizon: state.meta.max_forecast_horizon };
  }

  function scaleRow(rawRow, scale, min) {
    return rawRow.map((v, j) => v * scale[j] + min[j]);
  }

  async function recursiveForecast(startInput, horizon) {
    const { meta, scalers, model, elecByTs, weatherByTs, minTs, maxTs, minStart } = state;
    const startTs = toHourTs(startInput);

    if (startTs < minStart || startTs > maxTs) {
      throw new Error(`start must be between ${isoHour(minStart)} and ${isoHour(maxTs)}`);
    }
    const endTs = new Date(startTs.getTime() + hourMs(horizon - 1));
    if (endTs > maxTs) {
      const maxHorizonHere = Math.round((maxTs - startTs) / hourMs(1)) + 1;
      throw new Error(
        `horizon too large for this start time — historical data (used as the weather-forecast stand-in) ` +
        `only extends to ${isoHour(maxTs)}. Max horizon from this start is ${maxHorizonHere}h.`
      );
    }

    const predicted = new Map(); // iso -> value
    const featCache = new Map(); // iso -> raw feature array (ordered per meta.feature_cols)

    function elecValue(d) {
      const iso = isoHour(d);
      if (predicted.has(iso)) return predicted.get(iso);
      if (elecByTs.has(iso)) return elecByTs.get(iso);
      throw new Error(`missing electricity history at ${iso}`);
    }

    function computeFeatureRow(d) {
      const iso = isoHour(d);
      if (featCache.has(iso)) return featCache.get(iso);

      const w = weatherByTs.get(iso);
      if (!w) throw new Error(`missing weather history at ${iso}`);
      const tfeat = timeFeatures(d);
      const rowDict = { ...w, ...tfeat };

      for (const lag of meta.lag_hours) {
        const refD = new Date(d.getTime() - hourMs(lag));
        rowDict[`${meta.target_col}_lag_${lag}`] = elecValue(refD);
      }
      for (const win of meta.rolling_windows) {
        const vals = [];
        for (let k = 1; k <= win; k++) vals.push(elecValue(new Date(d.getTime() - hourMs(k))));
        const mean = vals.reduce((a, b) => a + b, 0) / vals.length;
        const variance = vals.reduce((a, b) => a + (b - mean) ** 2, 0) / (vals.length - 1); // ddof=1, matches pandas
        rowDict[`${meta.target_col}_roll_mean_${win}`] = mean;
        rowDict[`${meta.target_col}_roll_std_${win}`] = Math.sqrt(variance);
      }

      const row = meta.feature_cols.map((c) => rowDict[c]);
      featCache.set(iso, row);
      return row;
    }

    const results = [];
    for (let h = 0; h < horizon; h++) {
      const targetD = new Date(startTs.getTime() + hourMs(h));
      const windowRaw = [];
      for (let k = meta.lookback; k >= 1; k--) {
        windowRaw.push(computeFeatureRow(new Date(targetD.getTime() - hourMs(k))));
      }
      const windowScaled = windowRaw.map((row) => scaleRow(row, scalers.feature_scale, scalers.feature_min));

      const X = tf.tensor([windowScaled]); // [1, lookback, n_features]
      const yScaledTensor = model.predict(X);
      const yScaled = (await yScaledTensor.data())[0];
      X.dispose();
      yScaledTensor.dispose();

      const yPred = (yScaled - scalers.target_min[0]) / scalers.target_scale[0];
      const iso = isoHour(targetD);
      predicted.set(iso, yPred);

      const actual = elecByTs.has(iso) ? elecByTs.get(iso) : null;
      results.push({ timestamp: iso, predicted: Math.round(yPred * 10000) / 10000, actual });
    }

    return results;
  }

  return { init, getBounds, recursiveForecast, isoHour, toHourTs };
})();
